llama.cpp

```diff
@@ -3148,702 +3140,560 @@ static bool llm_load_tensors(
         void * progress_callback_user_data) {
     model.t_start_us = ggml_time_us();
 
-    auto & ctx     = model.ctx;
     auto & hparams = model.hparams;
 
+    model.main_gpu     = main_gpu;
     model.n_gpu_layers = n_gpu_layers;
 
-    size_t ctx_size = ggml_tensor_overhead() * ml.n_tensors;
+    size_t ctx_size = ggml_tensor_overhead()*ml.n_tensors;
+
+    // TODO: user configurable
+    enum gpu_split_mode {
+        CUDA_SPLIT_NONE,    // single GPU
+        CUDA_SPLIT_LAYER,   // offload layers to different GPUs
+        CUDA_SPLIT_ROW      // split matrix rows across GPUs
+    };
+
+    gpu_split_mode split_mode = CUDA_SPLIT_LAYER;
+    const int64_t n_layer     = hparams.n_layer;
+    const int64_t i_gpu_start = std::max((int64_t) hparams.n_layer - n_gpu_layers, (int64_t) 0);
+
 
-    LLAMA_LOG_INFO("%s: ggml ctx size       = %7.2f MiB\n", __func__, ctx_size/1024.0/1024.0);
+    // there is very little benefit to offloading the input layers, so always keep it on the CPU
+    model.buft_input = llama_default_buffer_type_cpu(true);
+    model.buft_layer.resize(n_layer);
+    // cpu layers
+    for (int64_t i = 0; i < i_gpu_start; ++i) {
+        model.buft_layer[i] = llama_default_buffer_type_cpu(true);
+    }
 
-    // create the ggml context
+#ifdef GGML_USE_CUBLAS
+    if (split_mode == CUDA_SPLIT_LAYER) {
+        int device_count = ggml_backend_cuda_get_device_count();
+        float splits[GGML_CUDA_MAX_DEVICES];
+        std::copy(tensor_split, tensor_split + device_count, splits);
+        bool all_zero = std::all_of(splits, splits + device_count, [](float x) { return x == 0.0f; });
+        if (all_zero) {
+            // set by free memory
+            for (int i = 0; i < device_count; ++i) {
+                size_t total;
+                size_t free;
+                ggml_backend_cuda_get_device_memory(i, &total, &free);
+                splits[i] = free;
+            }
+        }
+        float split_sum = 0.0f;
+        for (int i = 0; i < device_count; ++i) {
+            split_sum += splits[i];
+            splits[i] = split_sum;
+        }
+        for (int i = 0; i < device_count; ++i) {
+            splits[i] /= split_sum;
+            printf("split[%d] = %.2f\n", i, splits[i]);
+        }
+
+        int act_gpu_layers = std::min(n_gpu_layers, (int)n_layer + 1);
+
+        // assign layers proportionally, in reverse order
+        for (int64_t i = i_gpu_start; i < n_layer; ++i) {
+            int layer_gpu = std::upper_bound(splits, splits + device_count, float(i - i_gpu_start)/act_gpu_layers) - splits;
+            printf("layer %d -> gpu %d\n", (int)i, layer_gpu);
+            model.buft_layer[i] = llama_default_buffer_type_offload(layer_gpu);
+        }
+        // output layer
+        if (n_gpu_layers > n_layer) {
+            int layer_gpu = std::upper_bound(splits, splits + device_count, float(n_layer)/act_gpu_layers) - splits;
```

Comment:
Is there a scenario in which the non-repeating layers end up on any GPU other than the last one?

Reply:
Not currently, the `upper_bound` is not really necessary here.

Reply:
It is actually necessary because the last GPU(s) may have a fraction of 0, so we can't just use the last split for the last GPU.